test/ruby/test_sprintf.rb

```diff
@@ -507,6 +507,11 @@ def test_named_typed_enc
     end
   end
 
+  def test_coderange
+    str = "wrong constant name %s" % "\u3042"
+    refute_predicate str, :ascii_only?
+  end
```

Comment:
For only two lines, this test is rather tricky. You may want to comment it. As far as I can tell, you're relying on the string literal `"wrong constant name %s"` to have a populated code range of `ENC_CODERANGE_7BIT`. And you do need it to be populated, because if the code range were `ENC_CODERANGE_UNKNOWN`, then you'd end up doing a full code range scan regardless of the optimizations made in this PR.

Forcing the code range to be populated with `ascii_only?` to begin with might make it clearer:

```ruby
  def test_coderange
    # Ensure the string has an `ENC_CODERANGE_7BIT`  code range.
    str = "wrong constant name %s"
    assert str.ascii_only?

    # Ensure that the string's code range has changed.
    str = str % "\u3042"
    refute_predicate str, :ascii_only?
    
    # Optional check to rule out broken strings:
    # assert str.valid_encoding?
  end
```

And I'm not sure if you also want to ensure the resulting code range is `ENC_CODERANGE_VALID`, as this test would pass if it was `ENC_CODERANGE_BROKEN`.

Reply:
The problem is that there's no way to check the coderange from Ruby without loading an extension, which I don't think we should do from a `test/ruby/` file. I can make the test intent clearer though.